### ISSUE TITLE
Feature idea: a temporary execution barrier

### ISSUE THREAD

**Issue by marius311** [fonsp/Pluto.jl#298]
Hi guys, over the last few days, after having toyed with Pluto for a while, I spent a solid chunk of time using it for my actual real work. Aside from continuing to love reactivity and wondering how I've gone so long without it, I think I pinpointed one feature which I feel like could take my productivity while using Pluto to the next level, which I thought I'd propose (and apologies if this has already been discussed somewhere). 

For reference, I'm a physicist doing lots of free-form analysis work, probably pretty typical of your users in academia. Basically the thing I kept running into is I have a fairly "mature" notebook set up, and then I want to tweak something near the "beginning" of the DAG, but I might want to look at a few cells beyond that to check things are right, before letting the whole DAG run to the end, which may actually take minutes to tens of minutes in my case.

What I usually ended up doing is to just make the change, look at the "quick check" cells as the whole thing evaluates, and then just interrupt and try again if "things weren't right". Unfortunatey interrupt is not always reliable or even possible if external libraries are called, and somehow the whole thing felt frantic. 

So my idea is basically just to add a way to place a barrier in the DAG either before or after a given cell (I ran into cases where both would be useful). I made a mock-up of what the UI might look like below. The button could live alongside the hoverable add cell above/below button, and once clicked would remain persistent (and can be clicked again to turn off). To visually indicate the temperary break in reactivity, the UI should dim all the subsequent cells in the DAG and hide their output. It would be something like this:

![image](https://user-images.githubusercontent.com/1510968/90331497-b5a0dd00-df69-11ea-8b8a-63e238cfaff3.png)

In terms of what is saved in the notebook file, maybe the break in the DAG can be indicated by just commenting out code which is beyond the barrier, maybe with a `#= Pluto barrier ... =#` or something to avoid confusing with user code. 

I can imagine you might have some reservations about temporarily breaking reactivity, which I can definitely understand. I do think though that this isn't really getting rid of reactivity, instead its an optional "power" feature that kind of fits in the reactivity paradigm really well. If I recall too in your JuliaCon talk you mentioned work on a way to "check in" a bunch of changes at once. Maybe that's similar and you've already thought of this, it sounded like its trying to solve the same underlying problem, but if that's similar to "pausing" evaluation while editing several cells at once, I'll mention that the key thing here is that you get to look at the output of _some_ cells after you've made a change. 

Anyway, just an idea I had which I hope you guys can consider. I'm more than happy to discuss more or to be a guinea pig should you decide to purse something like this.

**Comment by fonsp** [comment]
Hi @marius311,

Thank you for the exciting suggestion! It sounds like you found this idea after some serious Pluto use, which makes it very valuable! I appreciate that you talked about the existing 'solutions' for this problem, and I can relate to your observations.

Your suggestion gave me an idea: you would have a "hotbar" in the corner of the screen with "filters" that you can drag onto your cells - this would be one of them. Other might be capturing terminal output, step-by-step debugging, memory profiling, a browser notification when the cell is done, saving cell output to a file, ...

Even if it is just this single filter (execution barrier), I think that it would create a friendlier learning curve to hide it in a corner, instead of showing it next to every cell. 

You talked about having the barrier both before and after a cell. I see why both would be useful, but the feature would be easier to understand if there was only one option - I would say _after_ the cell. What do you think?

Another comment was that we don't really need to save it to a file, right? It seems more like something you want during your writing session.

And blurring the cells is a very smart idea! I am worried that breaking the reactivity can turn your mental model into chaos, but this might solve it.

_A side note about stopping cells - of course it would be great if we can _fix_ the problems that it has. The implementation is somewhat improvised, but I don't have any good notebooks to test it on. So I'd love to hear more about what you found, and maybe even an example of a notebook where Stop is broken? (#228)_

Thanks again, look forward to hear from you.
-fons

**Comment by j-fu** [comment]
Hi I have the feeling that such a feature could provide what I like to have in #259, provided that such a barrier also becomes an element of the file format.

**Comment by ScottPJones** [comment]
For teaching and demos, what would be useful is maybe a button (like the present button, possibly using the present button), so that any cells that are past the current "slide" being presented are not yet evaluated.
When the notebook is loaded, it would be good to have some marker (stored in the notebook.jl file) to indicate how much of the notebook should be active.  Given that it's a useful technique to put cells at the end, that don't need to be displayed, but need to be executed, without the cell info messing up the presentation, there would also need to be a marker to indicate where the end of the inactive region is.
Maybe this could be done in some general fashion, being able to indicate blocks of cells, which are inactive initially, and can be activated as a set (such as when the presentation reaches that slide, or a button on that "slide")

In my case, it is to have an interactive demo, showing off the advantages of the package being demoed as well as Julia - and incidentally, showing off the coolness of Pluto compared to Jupyter!

**Comment by Moelf** [comment]
@marius311 I love this idea, clearly you know the pain point of users of "out kind". 

One comment: because of the reactivity model, we indeed have to do this "before" cell. Say we have a variable `params` that is used in `heaby_func`, you can't do "after the cell" without doing some cumbersome `real_paras=paras` in one cell and change another cell to `heaby_func(real_paras)` because we don't allow re-definition. 

In summary how this should work would be:
1. mark a cell as barrier
2. do the usual thing except for each cell parallel and after the barrier cell in DAG, don't run them
3. instead grey out these cells's output.

**Comment by pbouffard** [comment]
If this is implemented it would be great if Pluto would automatically flag certain cells as long-running (after they have first run and been observed as such) and ask the user the next time they are about to be run if they want to defer execution (with a "no, and never ask me again" option of course). Bonus: Pluto could offer to wrap the cell with some kind of memoization.
